arch/xtensa/src/common/xtensa_nputs.c

```diff
@@ -39,8 +40,14 @@
 
 void up_nputs(const char *str, size_t len)
```

Comment:
can `up_nputs` be used from the interrupt context?

Reply:
It isn't intended to...

Reply:
since up_putc can be called in the interrupt context, and up_nputs is an accelerated version of up_putc, it's better to make up_nputc work in the interrupt handler too.

Reply:
BTW, look like we need patch all arch up_nputs?

Reply:
> BTW, look like we need patch all arch up_nputs?

Yes, it's advisable to. Should we do this in this PR? (I can't test other architectures) or submit PRs specific to each arch?

Reply:
> since up_putc can be called in the interrupt context, and up_nputs is an accelerated version of up_putc, it's better to make up_nputc work in the interrupt handler too.

Do you mean `up_nputs`? I don't know if we should consider it to be used in the interrupt handler.
Instead of using a mutex (and making it work in ISRs), I can only think of protecting it with `spin_lock_irqsave`, but it'd deteriorate general performance.

Reply:
syslog can be called in interrupt context, we need some special handle like serial driver:
https://github.com/apache/nuttx/blob/master/drivers/serial/serial.c#L1189-L1209

Reply:
> syslog can be called in interrupt context, we need some special handle like serial driver: https://github.com/apache/nuttx/blob/master/drivers/serial/serial.c#L1189-L1209

Yes, but nowadays there are no problems calling syslog from ISR:

- if CONFIG_SYSLOG_INTBUFFER is set, then it'd be flushed outside the interrupt handler by [`syslog_write`](https://github.com/apache/incubator-nuttx/blob/2e54df0f35b743df26f5737aa6a9d7c74f12aa85/drivers/syslog/syslog_write.c#L140);
- Otherwise, the syslog will call the registered `sc_force` ([`syslog_default_putc`](syslog_default_putc) for the serial console). That would call `up_putc`.

Considering syslog, it wouldn't call `up_nputs` during the interrupt handler. Do you think we still need to provide some checks to avoid it being used directly from the ISR? (please consider that we already have https://github.com/apache/incubator-nuttx/blob/ef65d443ade0877f4b43a133e54565ba90640714/sched/semaphore/sem_wait.c#L80)

Reply:
> syslog can be called in interrupt context, we need some special handle like serial driver:
> https://github.com/apache/nuttx/blob/master/drivers/serial/serial.c#L1189-L1209

The family of low-level serial output functions is typically used (1) very early in initialization before the serial driver has been initialized and (2) in interrupt handlers.

Semaphores/mutexes should never be used before nxsem_initialize() is called.  Otherwise you would be using the semaphore subsystem before it has been initialized.

This may or may not work out for this case, but in general that is the architectural principle upon which the OS initialization was designed and should not be violated.

So either (1) up_nputs() must not use semaphores or (2) up_nputs() must not be used prior to calling nxsem_initialize().  There is a global variable g_nx_init_state that can also be used to check where nx_start is in the initialization sequence.

Reply:
> So either (1) up_nputs() must not use semaphores or (2) up_nputs() must not be used prior to calling nxsem_initialize(). There is a global variable g_nx_init_state that can also be used to check where nx_start is in the initialization sequence.

I don't think that it is a issue here, but one other thing that we should all keep in mind:  As a general rule, we should not use semaphores or mutexes on the IDLE thread.  The IDLE thread is the thread that runs when you boot up the system.  All initialization is performed on the IDLE thread, hence, it is sometimes called the initialization thread.

Why not?  Because of the result of calling sem_wait() were to cause the IDLE thread to attempt to wait, the system will crash.  The IDLE thread can never wait and must always be in the ready-to-run state.

This is not normally a problem during initialization because the system is running single-threaded during most of initialization and sem_wait() will never block while running single-threaded.

Reply:
Hi @patacongo !

Thank you for your observations. Indeed, if the `up_nputs` is meant to be used by the interrupt handler, it shouldn't use semaphores.

Reply:
@xiaoxiang781216 , I was thinking about https://github.com/apache/nuttx/pull/6618/commits/1a048f8ff703faa37e89f30f4775f32ef3829505 (from #6618): 

Could it be reverted (or a mutex be added there)? If syslog is called from the interrupt handler, it won't end up calling `syslog_default_write` (it'd use the registered `sc_force` instead, which is `syslog_default_putc`. By doing that, we can avoid creating a "thread-safe" version of `up_nputs`.

What do you think about it? It solves this same issue from an upper level.

Reply:
Ok, but it's better to use mutex_t instead semaphore

Reply:
PR updated!